### PR TITLE
Add perplexity metric to LanguageModel

### LINKED ISSUE #2400: LanguageModel should output perplexity
Is there a reason why `allennlp evaluate` doesn't already output the perplexity? This should be fairly easy to do, since you just take `exp(loss)`.

Seems like it might be tricky during training (@brendan-ai2 made a few notes about it in the code), but for evaluation we'd just need to override `get_metrics`, right?

### PR DESCRIPTION
**Pull request by thesamuel**
As discussed in #2400, this PR adds a `perplexity` metric to the `LanguageModel`. To do this, I created a `Perplexity` metric class which is updated with a negative log loss. @nelson-liu 

fixes #2400

### CODE REVIEW

File: allennlp/models/language_model.py [hunk 1]
```diff
@@ -300,8 +302,9 @@ def forward(self,  # type: ignore
                     average_loss = forward_loss / num_targets.float()
             else:
                 average_loss = torch.tensor(0.0).to(forward_targets.device)  # pylint: disable=not-callable
-            # this is stored to compute perplexity if needed
+            # TODO: _last_average_loss buffer is no longer used.
```
Comment (nelson-liu) [inline allennlp/models/language_model.py]: @brendan-ai2 , want to confirm that we're ok with just deleting `self._last_average_loss`?

File: allennlp/training/metrics/perplexity.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+@Metric.register("perplexity")
+class Perplexity(Average):
+    """
+    Perplexity is a common metric used for evaluating how well a language model
+    predicts a sample.
+
+    Notes
+    -----
+    Assumes negative log likelihood loss of each batch. Provides the average
```
Comment (nelson-liu) [inline allennlp/training/metrics/perplexity.py]: maybe specify that log should be base `e`?

File: allennlp/models/language_model.py (hunk 1 above)
Comment (thesamuel) [reply allennlp/models/language_model.py]: Removing this causes an error when restoring from a serialized model that was saved with this buffer. There is a test that fails when this is removed.

Comment (nelson-liu) [reply allennlp/models/language_model.py]: Yeah, it's a bit annoying---we either have to retrain a new fixture or go into the existing fixture and manually remove the `_last_average_loss` key in the state dictionary. (1) might be easier, and we probably already have a process to do it... @brendan-ai2 might know more?

File: allennlp/training/metrics/perplexity.py (hunk 2 above)
Comment (thesamuel) [reply allennlp/training/metrics/perplexity.py]:
```suggestion
    Assumes negative natural log likelihood loss of each batch. Provides the average
```

File: allennlp/models/language_model.py (hunk 1 above)
Comment (brendan-ai2) [reply allennlp/models/language_model.py]: The fact that the test with the fixture is breaking means that all serialized models would break if we remove it. This would be bad. Let's avoid that by just keeping the field around with a note that it's unused and exists for backwards compatibility.

File: allennlp/training/metrics/perplexity.py (hunk 2 above)
Comment (thesamuel) [reply allennlp/training/metrics/perplexity.py]: Fixed

File: allennlp/models/language_model.py (hunk 1 above)
Comment (thesamuel) [reply allennlp/models/language_model.py]: Sounds good. I added a comment above that buffer in `__init__` and removed the update from `forward`.

**Comment by thesamuel** [comment]
I'm unsure if my commit message is in the correct format. It seems like commits are automatically squashed on a per-PR basis, but I've been squashing manually.

**Comment by nelson-liu** [comment]
Don't worry about squashing, that's handled by GitHub when the PR is merged.

**Comment by matt-gardner** [comment]
@nelson-liu, @thesamuel, we're not sure what this is waiting on.  Do either of you know?

**Comment by thesamuel** [comment]
> @nelson-liu, @thesamuel, we're not sure what this is waiting on. Do either of you know?

As far as I know, there's nothing else left to resolve.

**Comment by thesamuel** [comment]
@nelson-liu Awesome, thanks! 

Also, could you double check whether I need to change my commit message? I tried to format it the way GitHub is going to, before I realized it was automatic.

**Comment by nelson-liu** [comment]
I think that GitHub generates the commit message from the PR title when squashing, so things should be fine. Thanks for thinking of it, though!
